### ISSUE TITLE
reader_concurrency_semaphore: admit new reads only when all current ones are stalled on I/O

### ISSUE THREAD

**Issue by denesb** [scylladb/scylla#5718]
From @avikivity:
> use the tracking to distinguish between a stalled and an unstalled reader, and only admit new readers if all readers are stalled.
>
> A reader is stalled if it cannot consume cpu to make progress:
>  - it is waiting for I/O to complete>
>  - it is waiting for someone to call flat_mutation_reader::fill_buffer>
>  - some other read-specific stalling (like multishard_mutation_reader waiting for data from other shards, or a combining reader waiting for its sub-readers)
>
> An unstalled reader is one that could make progress if you just gave it some cpu.
>
> A similar concept is voluntary and involuntary context switches for a unix process. A voluntary context switch is one that the process initiated by doing some I/O or waiting for a mutex. An involuntary context switch is caused by the kernel preempting the process and giving 
the cpu to someone else.
>
> The idea is to change flat_mutation_reader, tracked_file, and individual mutation_reader implementations that can wait to record their stall/unstall state together with their memory consumption. Then, we change the semaphore not to admit new reads while there is an unstalled 
reader. It's better to let the unstalled reader to make as much progress as it can, since that both reduces transient memory usage, and also increases the accuracy of memory tracking since it will call consume() as it allocates more resources. Having unstalled readers compete for CPU 
just makes everyone slower and consumes more memory.

**Comment by denesb** [comment]
Refs: #4077.

**Comment by slivne** [comment]
This seems correct in the case that we have memory limit / the disks are not the bottleneck. 

In case the disk is the bottleneck and not memory or cpu I think this will increase the latency - we are not maxing out the disk capacity.

**Comment by avikivity** [comment]
Why? As soon as a read becomes a miss, we will start a new one.

**Comment by avikivity** [comment]
Here's a simple way to do it: we already have tracking_file_impl talking to a reader_permit. Extend it to count I/O concurrency. When a reader_permit's outstanding I/O count becomes nonzero, it is I/O bound. When it moves back to zero, it becomes CPU bound.

We can now admit reads to the cache only when there are no CPU bound queries.

Note concurrency accounting should be done when the read begins/ends, not when the buffer is released, since a buffer can long outlive I/O.

**Comment by avikivity** [comment]
I don't think my idea of using tracking_file_impl works. First, it doesn't handle read-aheads. We can have pending I/O that is not blocking anyone.

**Comment by avikivity** [comment]
It also doesn't handle multishard readers, where the files are remote.

**Comment by avikivity** [comment]
So we'll have to mark blockages manually.

1. change flat_mutation_reader::impl c'tor/d'tor to count the number of impl:s serving this permit
2. every place we wait for non-local-cpu work (sstable index reads, sstable data reads, multishard foreign reads) we bump another counter in the permit. Note sometimes we wait for two operations in parallel (index start and end reads), we have to adjust for that
3. if the number of wait regions is equal to the number of impls for the permit, then that permit is blocked and we should allow another one in

This means more manual tagging, but it's also more accurate in that if *any* impl is able to consume cpu, we let it, rather than increasing concurrency.

**Comment by denesb** [comment]
I don't think the condition should be count >= count(impl). We should treat the reader pipeline as a tree. Blocked should propagate up the tree. E.g. a combined reader is only blocked if at least one of its subreaders are blocked. Similarly in the case of the multishard reader, if the current shard is blocked, the reader is blocked despite all the non-current shard readers being available. We propagate this up and if the blocked status reaches the top level reader (restricted reader) we signal the semaphore.

**Comment by avikivity** [comment]
I think you are correct. If you can reuse the existing tree structure, I think it's worthwhile. If we have to create a parallel structure, it's enough to use any(non-blocked) -> this tree is non-blocked, on the assumption that pretty soon all the other branches will block too.

**Comment by nyh** [comment]
Similar to @slivne's comment above, I don't understand the *goal* of reader concurrency semaphore after your change, and this goal should be explicitly stated, not just implied amidst lengthy discussions on the implementation details like you and Avi seem to be doing today...

I **thought** that the goal of reader concurrency semaphore was to to limit the amount of memory that the readers use. However, now you are suggesting that even if we're at that memory limit, if none of these readers want any CPU, we can allow more readers. Maybe this makes sense for various reasons - but it means the memory limit is no longer an actual memory limit. So it begs the question, what is the new goal of the reader concurrency semaphore? Why do we even have it?

**Comment by denesb** [comment]
The task of the reader concurrency semaphore is to regulate access to system resources for reads. This has two parts:
* Limit the amount of memory they are allowed to consume, and
* Make sure the resources are well used.
This latter part is where Avi's suggestions come in. For example, even if we omit the memory consumption completely, it still doesn't make sense to admit infinite amount of reads. They will start firing infinite amount of I/O and none of them can complete in time. It makes more sense to admit just a subset, that don't oversaturate the I/O subsystem and therefore can make good progress. Avi's suggestions extends this to the CPU. Strictly speaking, from the point of view of the CPU it only makes sense to admit a single read. After all in seastar's sharded architecture a single shard really only works on a single task at a time. Even if the read is preemted, if all it needs is more CPU it makes more sense to continue it later, rather than start a new read. So we want to admit new reads only when the existing reads cannot make more progress given more CPU time, if they are e.g. waiting on I/O or another shard.

**Comment by avikivity** [comment]
The goal has not changed.

The condition is inverted, instead of
> if we're at that memory limit, if none of these readers want any CPU, we can allow more readers.
It's really
> if we're below the memory limit, if any reader wants CPU, don't allow more readers.

The reason is that we can't predict how much memory readers will consume in practice. They start consuming memory as they run. When they start, they consume very little memory. By letting a ready-to-run reader to proceed, instead of preempting it with another reader, we let it claim what memory it wants, or complete if it doesn't need to do any I/O.

It's also better to restrict concurrency for other reasons - we check for timeouts when we admit, so the later we admit, the more chances we have to drop the query without starting to execute it.

**Comment by nyh** [comment]
> The task of the reader concurrency semaphore is to regulate access to system resources for reads. This has two parts:
> 
>     * Limit the amount of memory they are allowed to consume, and
> 
>     * Make sure the resources are well used.
> 
> 
> This latter part is where Avi's suggestions come in. For example, even if we omit the memory consumption completely, it still doesn't make sense to admit infinite amount of reads. They will start firing infinite amount of I/O and none of them can complete in time.

This doesn't convince me. We could have (or already have? is all of this documented anywhere?) a separate semaphore which limits the number of really active disk reads - there is no point of doing 1000 reads in parallel if the physical disk can only do 10 (or even 1, considering that long reads can be internally parallelized). You don't have to do it in the admittance level.

> It makes more sense to admit just a subset, that don't oversaturate the I/O subsystem and therefore can make good progress. Avi's suggestions extends this to the CPU. Strictly speaking, from the point of view of the CPU it only makes sense to admit a single read. After all in seastar's sharded architecture a single shard really only works on a single task at a time. Even if the read is preemted, if all it needs is more CPU it makes more sense to continue it later, rather than start a new read. So we want to admit new reads only when the existing reads cannot make more progress given more CPU time, if they are e.g. waiting on I/O or another shard.

Maybe this makes sense, maybe it doesn't. Here is an example why it doesn't: Consider that we already have 10 ongoing reads, all waiting for the same slow disk. We can admit another read, but all that will happen is that it too will soon reach a point that it too will just wait for the same slow disk. What bothers me more, however, is that it looks to me like a new feature that it isn't clear how it's at all, if at all, related to the memory limit.

**Comment by denesb** [comment]
> > The task of the reader concurrency semaphore is to regulate access to system resources for reads. This has two parts:
> > ```
> > * Limit the amount of memory they are allowed to consume, and
> > 
> > * Make sure the resources are well used.
> > ```
> > 
> > 
> > This latter part is where Avi's suggestions come in. For example, even if we omit the memory consumption completely, it still doesn't make sense to admit infinite amount of reads. They will start firing infinite amount of I/O and none of them can complete in time.
> 
> This doesn't convince me. We could have (or already have? is all of this documented anywhere?) a separate semaphore which limits the number of really active disk reads - there is no point of doing 1000 reads in parallel if the physical disk can only do 10 (or even 1, considering that long reads can be internally parallelized). You don't have to do it in the admittance level.

Doesn't convince you of what? That this feature is needed at all? That it should be in the existing `reader_concurrency_semaphore`? Something else?

Admitting new reads only when all current ones are waiting on I/O doesn't contradict what you are saying, and it certainly doesn't prevent us from implementing restriction on I/O concurrency. But that is not what this issue is about.

> 
> > It makes more sense to admit just a subset, that don't oversaturate the I/O subsystem and therefore can make good progress. Avi's suggestions extends this to the CPU. Strictly speaking, from the point of view of the CPU it only makes sense to admit a single read. After all in seastar's sharded architecture a single shard really only works on a single task at a time. Even if the read is preemted, if all it needs is more CPU it makes more sense to continue it later, rather than start a new read. So we want to admit new reads only when the existing reads cannot make more progress given more CPU time, if they are e.g. waiting on I/O or another shard.
> 
> Maybe this makes sense, maybe it doesn't. Here is an example why it doesn't: Consider that we already have 10 ongoing reads, all waiting for the same slow disk. We can admit another read, but all that will happen is that it too will soon reach a point that it too will just wait for the same slow disk. What bothers me more, however, is that it looks to me like a new feature that it isn't clear how it's at all, if at all, related to the memory limit.

In short, it doesn't, it relates to maintaining a read concurrency such that the resources are used with minimal waste. The resource this issue focuses on is the CPU, it doesn't consider the disk at all and implementing this feature shouldn't make I/O any worse (we should measure of course).

I am too uncertain of the exact effects of this feature so we should definitely do a POC before commiting to it.

**Comment by avikivity** [comment]
Let's define reader_concurrency_semaphore's mission as "keep concurrency as low as possible, not not lower".

Why is low concurrency important?
 - High concurrency results in high memory consumption. Even before OOM, that comes at the expense of cache.
 - High concurrency increases the average latency of admitted requests. This is bad, because the average latency of successful requests is at least the average latency of admitted requests. If we reduce the average latency of admitted requests, we have the potential to reduce the average latency of successful requests (by refusing to admit a larger proportion of all requests)

On the other hand, too low concurrency is also bad:
 - below 1 running request per shard, we have idle CPU
 - before <disk-dependent> blocked requests per shard, we have less than fully-utilized disks

The conclusion is that we should admit requests until we know it won't help any more (CPU or disk saturation) or will hurt (running out of memory). CPU saturation is easy (1 per shard). Disk saturation is harder, we currently hardcode 100 per shard.
